Support ujson.loads(bytearray(...)) and other bytes-like objects.

bwoodsend opened the pull request:
Fixes #572

Changes proposed in this pull request:

* Generalise the support for using `ujson.loads()` on bytes to any C contiguous bytes like object such as `bytearray()`, `memoryview()` and `array.array()`.

bwoodsend committed:
Support ujson.loads(bytearray(...)) and other bytes-like objects.

bwoodsend commented:
Ughh, I hate supporting PyPy

JustAnotherArchivist commented:
Looks like this upstream issue might be related: https://foss.heptapod.net/pypy/pypy/-/issues/3520

bwoodsend commented:
Hmm, I'd have expected this line to fail too if it was the same cause since `bytes()` would have returned the original unsliced bytes string:

https://github.com/ultrajson/ultrajson/blob/08099408c9b0c2cbe541c8ad6ea0a77493546cb3/tests/test_ujson.py#L1156

I'm just going to create a new ticket.

bwoodsend commented:
Filed upstream: https://foss.heptapod.net/pypy/pypy/-/issues/3872

How do we want to handle this?

1. Merging this PR without adding some kind of additional safeguard for PyPy sounds to me like another CVE in the making.
2. Cancelling this PR just because of PyPy feels silly.
3. Adding *if PyPy: use old logic / else use new logic* is both ugly from the developer's perspective (all the fiddly *remember to dec-ref/free this object in all successful and erroneous cases* would double) and potentially confusing to users who may wonder why ujson happily reads `bytearray()` under CPython but not PyPy.
4. Any better ideas?

JustAnotherArchivist commented:
How about an extra test inside the bytes-like case for PyPy, testing specifically for `bytes` and `bytearray`? That keeps the error condition code minimal, and it'd still add support for `bytearray` on PyPy. We basically have to exclude anything that isn't guaranteed to be C-contiguous (which `bytes` and `bytearray` are, I think, although I doubt it's documented anywhere) until that bug is fixed.

```c
  bool is_bytes_like = !PyObject_GetBuffer(arg, &buffer, PyBUF_C_CONTIGUOUS);
  if (is_bytes_like)
  {
    #ifdef PYPY_VERSION
      // PyPy's buffer protocol implementation is buggy: https://foss.heptapod.net/pypy/pypy/-/issues/3872
      if (!PyBytes_Check(arg) && !PyByteArray_Check(arg)) {
        PyBuffer_Release(&buffer);
        PyErr_Format(PyExc_TypeError, "Expected string, bytes, or bytearray object");
        return NULL;
      }
    #endif
    raw = buffer.buf;
    sarg_length = buffer.len;
  }
```

Not sure what we'd want to do once the bug is fixed in PyPy though. Maybe we could change that to a version check, either in the code or somewhere in the build process, but not sure that's the best approach.

bwoodsend committed:
Limit loading from bytes-like to just bytes() and bytearray() on PyPy.

Due to a bug in PyPy [1], PyObject_GetBuffer() fails to detect non C
contiguous inputs. Allowing ujson.loads() to run on non-contiguous
buffers such as a strided NumPy array would lead to mayhem. Approximate
the intention of PyObject_GetBuffer() by explicitly checking that the
input is within the set of common bytes-like objects known to be
unconditionally C contiguous.

- [1]: https://foss.heptapod.net/pypy/pypy/-/issues/3872

JustAnotherArchivist commented on python/JSONtoObj.c:
```diff
@@ -185,40 +187,60 @@ PyObject* JSONToObj(PyObject* self, PyObject *args, PyObject *kwargs)
+    else
     {
-      //Exception raised above us by codec according to docs
+      PyErr_Format(PyExc_TypeError, "Expected String or C contiguous Bytes-like object");
```
The capitalisation (which I know predates this change) is odd. But more importantly, I think we should add another PyPy check here so the error messages aren't confusing under PyPy when passing a bytes-like object that isn't `bytes` or `bytearray` vs passing something completely incompatible like an `int`.

```suggestion
      #ifdef PYPY_VERSION
        PyErr_Format(PyExc_TypeError, "Expected string, bytes, or bytearray");
      #else
        PyErr_Format(PyExc_TypeError, "Expected string or C-contiguous bytes-like object");
      #endif
```

JustAnotherArchivist commented on python/JSONtoObj.c:
```diff
@@ -185,40 +187,60 @@ PyObject* JSONToObj(PyObject* self, PyObject *args, PyObject *kwargs)
-  {
-    sarg = PyUnicode_AsEncodedString(arg, NULL, "surrogatepass");
-    if (sarg == NULL)
+  else {
```
Nit: the bracket's on a separate line in almost the entire code.

bwoodsend replied:
Grr, well spotted. How I hate that convention - such a waste of a line.

JustAnotherArchivist commented on python/JSONtoObj.c:
```diff
@@ -185,40 +187,60 @@ PyObject* JSONToObj(PyObject* self, PyObject *args, PyObject *kwargs)
   {
     Py_DECREF(sarg);
   }
+  else {
```
Same nit as above: bracket should be on a separate line.

JustAnotherArchivist reviewed (commented):
Some minor things, else LGTM.
